app/assets/javascripts/discourse/app/lib/textarea-text-manipulation.js

```diff
@@ -285,8 +285,8 @@ export default class TextareaTextManipulation {
 
         if (
           operation !== OP.ADDED &&
-          ((l.slice(0, hlen) === hval && tlen === 0) ||
-            (tail.length && l.slice(-tlen) === tail))
+          l.slice(0, hlen) === hval &&
+          (tlen === 0 || l.slice(-tlen) === tail)
```

Comment:
Additional information: `tlen` here is `tail.length` (https://github.com/Lhcfl/discourse/blob/fix-applysurround/app/assets/javascripts/discourse/app/lib/textarea-text-manipulation.js#L214), so the original code only determines the two branches based on the difference of `tlen`